site/source/docs/api_reference/wasm_workers.rst

```diff
@@ -0,0 +1,355 @@
+hierarchy manually by posting the Worker creation over to the main thread yourself.
+
+Note that support for nested Workers varies across browsers. As of 02/2022, nested Workers are `not
+supported in Safari <https://bugs.webkit.org/show_bug.cgi?id=22723>`_. See `here 
+<https://github.com/johanholmerin/nested-worker>`_ for a polyfill.
+
+Pthreads can use the Wasm Worker synchronization API, but not vice versa
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The multithreading synchronization primitives offered in ``emscripten/wasm_worker.h``
+(``emscripten_lock_*``, ``emscripten_semaphore_*``, ``emscripten_condvar_*``) can be freely invoked
+from within pthreads if one so wishes, but Wasm Workers cannot utilize any of the synchronization
+functionality in the Pthread API (``pthread_mutex_*``, ``pthread_cond_``, ``pthread_rwlock_*``, etc),
+since they lack the needed pthread runtime.
+
+Pthreads have a "thread main" function and atexit handlers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The startup/execution model of pthreads is to start up executing a given thread entry point function.
+When that function exits, the pthread will also (by default) quit, and the Worker hosting that pthread
+will return to the Worker pool to wait for another thread to be created on it.
+
+Wasm Workers instead implement the direct web-like model, where a newly created Worker sits idle in its
+event loop, waiting for functions to be posted to it. When those functions finish, the Worker will
+return to its event loop, waiting to receive more functions (or worker scope web events) to execute.
+A Wasm Worker will only quit with a call to ``emscripten_terminate_wasm_worker(worker_id)`` or
+``emscripten_terminate_all_wasm_workers()``.
+
+Pthreads allow one to register thread exit handlers via ``pthread_atexit``, which will be called when
+the thread quits. Wasm Workers do not have this concept.
+
+Pthreads have a per-thread incoming proxy message queue, Wasm Workers do not
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In order to enable flexible synchronous execution of code on other threads, and to implement support
+APIs for example for MEMFS filesystem and Offscreen Framebuffer (WebGL emulated from a Worker) features,
+main browser thread and each pthread have a system-backed "proxy message queue" to receive messages.
+
+This enables user code to call API functions ``emscripten_sync_run_in_main_thread*()``,
+``emscripten_sync_run_in_main_runtime_thread()``, ``emscripten_async_run_in_main_runtime_thread()``,
+``emscripten_dispatch_to_thread()``, etc. from ``emscripten/threading.h`` to perform proxied calls.
+
+Wasm Workers do not provide this functionality. If needed, such messaging should be implemented manually
+by users via regular multithreaded synchronized programming techniques (mutexes, futexes, semaphores, etc.)
```

Comment:
What would it take to make the proxying.h API work with Wasm Workers? Is it just a matter of using something more general than pthread_t as the internal "thread" identifier? Do Wasm Workers have something similar to `pthread_self()`?

Reply:
Wasm Workers have `emscripten_wasm_worker_self_id()` for the same purpose as `pthread_self()`. I presume that proxying.h would work for Wasm Workers, it just has to properly DCE out if not used by the calling program.